fix(core): move generated i18n statements to the `consts` field of ComponentDef

AndrewKushnir opened the pull request:
This commit updates the code to move generated i18n statements into the `consts` field of
ComponentDef to avoid invoking `$localize` function before component initialization (to better
support runtime translations) and also avoid problems with lazy-loading when i18n defs may not
be present in a chunk where it's referenced.

Fixes #37092

Prior to this change the i18n statements were generated at the top leve:

```
var I18N_0;
if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
    var MSG_X = goog.getMsg(“…”);
    I18N_0 = MSG_X;
} else {
    I18N_0 = $localize('...');
}

defineComponent({
    // ...
    template: function App_Template(rf, ctx) {
        i0.ɵɵi18n(2, I18N_0);
    }
});
```

This commit updates the logic to generate the following code instead:

```
defineComponent({
    // ...
    consts: function() {
        var I18N_0;
        if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
            var MSG_X = goog.getMsg(“…”);
            I18N_0 = MSG_X;
        } else {
            I18N_0 = $localize('...');
        }
        return [
            I18N_0
        ];
    },
    template: function App_Template(rf, ctx) {
        i0.ɵɵi18n(2, 0);
    }
});
```

Also note that i18n template instructions now refer to the `consts` array using an index (similar to other template instructions).


## PR Type
What kind of change does this PR introduce?

- [x] Bugfix


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
fix(core): move generated i18n statements to the `consts` field of ComponentDef

This commit updates the code to move generated i18n statements into the `consts` field of
ComponentDef to avoid invoking `$localize` function before component initialization (to better
support runtime translations) and also avoid problems with lazy-loading when i18n defs may not
be present in a chunk where it's referenced.

Prior to this change the i18n statements were generated at the top leve:

```
var I18N_0;
if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
    var MSG_X = goog.getMsg(“…”);
    I18N_0 = MSG_X;
} else {
    I18N_0 = $localize('...');
}

defineComponent({
    // ...
    template: function App_Template(rf, ctx) {
        i0.ɵɵi18n(2, I18N_0);
    }
});
```

This commit updates the logic to generate the following code instead:

```
defineComponent({
    // ...
    consts: function() {
        var I18N_0;
        if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
            var MSG_X = goog.getMsg(“…”);
            I18N_0 = MSG_X;
        } else {
            I18N_0 = $localize('...');
        }
        return [
            I18N_0
        ];
    },
    template: function App_Template(rf, ctx) {
        i0.ɵɵi18n(2, 0);
    }
});
```

Also note that i18n template instructions now refer to the `consts` array using an index
(similar to other template instructions).

AndrewKushnir commented:
[Presubmit](http://test/OCL:325933368:BASE:325933383:1597110440114:98f4bf42)

AndrewKushnir commented:
Note: while having `MSG_X` vars inside other functions works correctly in Closure, it looks like Closure now requires `I18N_X` vars to be unique as well (similar to `MSG_X` ones), so some additional changes are needed to support that.

petebacondarwin commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -2006,30 +2006,32 @@ describe('i18n support in the template compiler', () => {
     });
 
-    it('should be generated for ICU-only i18n blocks', () => {
+    fit('should be generated for ICU-only i18n blocks', () => {
```
👀

petebacondarwin commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -197,9 +197,13 @@ export function compileComponentFromMetadata(
   definitionMap.set('vars', o.literal(templateBuilder.getVarCount()));
 
   // e.g. `consts: [['one', 'two'], ['three', 'four']]
```
Does this comment need to change to also illustrate the function form?

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -103,6 +103,15 @@ export function prepareEventListenerParameters(
+  extraStatements: o.Statement[];
+}
+
+function initComponentDefConsts(): ComponentDefConsts {
```
NIT - I would call this `createComponentDefConsts()`.

petebacondarwin reviewed (commented):
Looking good so far. Thanks for getting this moving @AndrewKushnir

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:326275924:BASE:326577982:1597374239942:6a8788ec).

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -304,15 +317,19 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
     this._bindingScope.notifyImplicitReceiverUse();
   }
 
+  private i18nGenerateMainBlockVar(): o.ReadVarExpr {
```
NIT - a short JSDoc comment would help here. I didn't understand what "main block variable" meant on first reading. What is the reason for the i18n prefix here? From reading the context I would have expected a method called `generateI18nVariable`.


Oh! I see that you were being consistent with the `i18nGenerateClosureVar()` method, which I think is equally poorly named. If you want avoid too much churn here, I am happy for your to leave this as-is but just add a jsdoc, please.

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -103,6 +103,18 @@ export function prepareEventListenerParameters(
+// When a constant requires some pre-processing, the `extraStatements` section contains
+// corresponding statements.
+export interface ComponentDefConsts {
+  extraStatements: o.Statement[];
+  returnList: o.Expression[];
```
I am fairly happy with these property names, but since you asked if I had other ideas, how about:

```ts
{ prepareStatements, constExpressions }
```

or

```ts
{ factories, constants }
```

???

petebacondarwin reviewed (approved):
LGTM! Nice work @AndrewKushnir.
Looks like you need to fix up some of the tests in `i18n_spec.ts` to account for the parameter change in `ɵɵi18nStart()` and `ɵɵi18nAttributes()`.

AndrewKushnir committed:
refactor(compiler): i18n compiler tests refactoring

This commit refactors i18n compiler tests to avoid code duplication and simplify further maintenance and updates.

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:326275924:BASE:326783463:1597468572564:5c96aa88).

petebacondarwin commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -182,6 +182,104 @@ const verify = (input: string, output: string, extra: any = {}): void => {
+// Describes a simple key-value object.
+type KVList = {
+  [key: string]: string
+};
```
Same as `Record<string, string>` no?

petebacondarwin replied:
Actually given that this is only used for `meta` then perhaps it could be an interface?

```
interface Meta {
  desc?: string;
  meaning?: string;
  id?: string;
}
```

AndrewKushnir replied:
Thanks for catching this Pete!

petebacondarwin commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -182,6 +182,104 @@ const verify = (input: string, output: string, extra: any = {}): void => {
+// Describes placeholder type used in tests. Note: the type is an array (not an object), since it's
+// important to preserve the order of placeholders (so that we can compare it with generated
+// output).
+type Placeholder = string[];
```
```suggestion
type Placeholder = [string, string];
```

petebacondarwin commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -182,6 +182,104 @@ const verify = (input: string, output: string, extra: any = {}): void => {
+
+// Generates a string that represents expected i18n block content for an ICU.
+const i18nIcuMsg =
+    (message: string, placeholders: string[][] = []) => {
```
```suggestion
    (message: string, placeholders: Placeholder[] = []) => {
```

petebacondarwin commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -643,4 +635,4 @@ describe('Runtime i18n', () => {
     });
   });
-});
+});
```
😱

petebacondarwin reviewed (approved):
A couple of minor nits - but LGTM!
Nice refactoring of the tests.

AndrewKushnir commented:
Note to Caretaker: this change should be safe (since runtime translations are not yet a public API), but we (Pete and myself) discussed this and agreed to make it `master-only` to avoid any risks for the stable branch (since the `$localize` function calls are now invoked at a different time). Thank you.
